src/ua_types_encoding_binary.c

```diff
@@ -1718,6 +1914,95 @@ calcSizeBinaryStructure(const void *p, const UA_DataType *type) {
     return s;
 }
 
+static size_t
+calcSizeBinaryStructureWithOptFields(const void *p, const UA_DataType *type) {
+    size_t s = 0;
+    size_t optFieldsSize = getOptFieldsSize(type);
+    uintptr_t boolptr = (uintptr_t)p;
+    uintptr_t ptr = boolptr + (optFieldsSize * sizeof(UA_Boolean));
+    u8 membersSize = type->membersSize;
+    const UA_DataType *typelists[2] = { UA_TYPES, &type[-type->typeIndex] };
+
+    /* Adding the size of the encoding mask */
+    s += sizeof(UA_UInt32);
+
+    /* Loop over members */
+    for(size_t i = 0; i < membersSize; ++i) {
+        const UA_DataTypeMember *member = &type->members[i];
+        const UA_DataType *membertype = &typelists[!member->namespaceZero][member->memberTypeIndex];
+        ptr += member->padding;
+
+        if(member->isOptional){
+            UA_Boolean fieldDefined = *(UA_Boolean*)boolptr;
+            boolptr += sizeof(UA_Boolean);
+            if(!fieldDefined){
+                if(member->isArray){
+                    ptr += sizeof(size_t) + sizeof(void*);
+                }
+                else{
+                    ptr += membertype->memSize;
+                }
+                continue;
+            }
+        }
+        /* Array */
+        if(member->isArray) {
+            const size_t length = *((const size_t*)ptr);
+            ptr += sizeof(size_t);
+            s += Array_calcSizeBinary(*(void *UA_RESTRICT const *)ptr, length, membertype);
+            ptr += sizeof(void*);
+            continue;
+        }
+
+        /* Scalar */
+        s += calcSizeBinaryJumpTable[membertype->typeKind]((const void*)ptr, membertype);
+        ptr += membertype->memSize;
+    }
+
+    return s;
+}
+
+static size_t
```

Comment:
This function is defined in multiple places.
It could be made non-static with a single implementation.